pkg/lifecycle/service.go

```diff
@@ -593,69 +666,77 @@ func (s *Service) runPipeline(ctx context.Context, pl *Instance) error {
 		})
 	}
 
-	s.updateOldStatusMetrics(pl)
-
-	pl.SetStatus(StatusRunning)
-	pl.Error = ""
-
-	s.updateNewStatusMetrics(pl)
-
-	err := s.store.Set(ctx, pl.ID, pl)
+	err := s.pipelines.UpdateStatus(ctx, rp.pipeline.ID, pipeline.StatusRunning, "")
 	if err != nil {
-		return cerrors.Errorf("pipeline not updated: %w", err)
+		return err
 	}
 
 	// cleanup function updates the metrics and pipeline status once all nodes
 	// stop running
-	pl.t.Go(func() error {
+	rp.t.Go(func() error {
 		// use fresh context for cleanup function, otherwise the updated status
 		// won't be stored
 		ctx := context.Background()
 
 		nodesWg.Wait()
-		err := pl.t.Err()
-
-		s.updateOldStatusMetrics(pl)
+		err := rp.t.Err()
 
 		switch err {
 		case tomb.ErrStillAlive:
 			// not an actual error, the pipeline stopped gracefully
-			err = nil
 			if isGracefulShutdown.Load() {
 				// it was triggered by a graceful shutdown of Conduit
-				pl.SetStatus(StatusSystemStopped)
+				err = s.pipelines.UpdateStatus(ctx, rp.pipeline.ID, pipeline.StatusSystemStopped, "")
 			} else {
 				// it was manually triggered by a user
-				pl.SetStatus(StatusUserStopped)
+				err = s.pipelines.UpdateStatus(ctx, rp.pipeline.ID, pipeline.StatusUserStopped, "")
+			}
+			if err != nil {
+				return err
 			}
 		default:
 			if cerrors.IsFatalError(err) {
-				pl.SetStatus(StatusDegraded)
 				// we use %+v to get the stack trace too
-				pl.Error = fmt.Sprintf("%+v", err)
+				err = s.pipelines.UpdateStatus(ctx, rp.pipeline.ID, pipeline.StatusDegraded, fmt.Sprintf("%+v", err))
+				if err != nil {
+					return err
+				}
 			} else {
-				pl.SetStatus(StatusRecovering)
-				// TODO: Implement backoff strategy
+				err = s.pipelines.UpdateStatus(ctx, rp.pipeline.ID, pipeline.StatusRecovering, "")
+				if err != nil {
+					return err
+				}
 			}
 		}
 
 		s.logger.
 			Err(ctx, err).
-			Str(log.PipelineIDField, pl.ID).
+			Str(log.PipelineIDField, rp.pipeline.ID).
 			Msg("pipeline stopped")
 
-		s.notify(pl.ID, err)
+		// confirmed that all nodes stopped, we can now remove the pipeline from the running pipelines
+		s.m.Lock()
+		delete(s.runningPipelines, rp.pipeline.ID)
+		s.m.Unlock()
+
+		s.notify(rp.pipeline.ID, err)
 		// It's important to update the metrics before we handle the error from s.Store.Set() (if any),
 		// since the source of the truth is the actual pipeline (stored in memory).
-		s.updateNewStatusMetrics(pl)
-
-		storeErr := s.store.Set(ctx, pl.ID, pl)
-		if storeErr != nil {
-			return cerrors.Errorf("pipeline not updated: %w", storeErr)
-		}
-
-		return err
+		return s.pipelines.UpdateStatus(ctx, rp.pipeline.ID, rp.pipeline.GetStatus(), "")
```

Comment:
This call is redundant now, since we call `UpdateStatus` above with the corresponding status, we can simply return `err` here.

Reply:
Good catch.